server/src/main/java/org/opensearch/action/search/QueryLatencyInsightService.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.search;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.Map;
+import java.util.concurrent.PriorityBlockingQueue;
+
+/**
+ * Service responsible for gathering, analyzing, storing and exporting
+ * top N queries with high latency data for search queries
+ *
+ * @opensearch.internal
+ */
+public class QueryLatencyInsightService extends QueryInsightService<
```

Comment:
should we make it abstract and maybe call it TopNQueryInsightService? Latency is just a metric/record type in this case, thus a service shouldn't be named after it. Also, this service isn't contain any latency related logic.

Reply:
Good point, let me refactor this.